src/win/wwindow.c

```diff
@@ -731,7 +732,13 @@ static LRESULT CALLBACK window_callback(HWND hWnd, UINT message,
          int vcode = wParam;
          int scode = (lParam >> 16) & 0xff;
          bool extended = (lParam >> 24) & 0x1;
-         _al_win_kbd_handle_key_release(scode, vcode, extended, win_display);
+         bool previous = (lParam >> 30) & 0x1;
+
+         /* The docs say that previous should always be 1, but it's not in practice.
+          * The events with previous = 0 seem malformed? I can get them reliably by
+          * pressing + holding Shift and then tapping Alt. */
+         if (previous)
+            _al_win_kbd_handle_key_release(scode, vcode, extended, win_display);
```

Comment:
I couldn't find anything suggesting that `previous == 0` is possible, very strange. Just ignoring these bad events seems alright.